GVFS/GVFS.Service.UI/WinToastNotifier.cs

```diff
@@ -0,0 +1,102 @@
+
+        public WinToastNotifier(string appId, ITracer tracer)
+        {
+            this.appId = appId;
+            this.tracer = tracer;
+        }
+
+        public ToastCallback UserResponseCallback { get; set; }
+
+        public void Notify(string title, string message, string actionButtonTitle, string callbackArgs)
+        {
+            ToastData toastData = new ToastData();
+
+            toastData.Visual = new VisualData();
+
+            BindingData binding = new BindingData();
+            toastData.Visual.Binding = binding;
+
+            binding.Template = "ToastGeneric";
+            binding.Items = new XmlList<BindingItem>();
+            binding.Items.Add(new BindingItem.TextData(title));
+            binding.Items.Add(new BindingItem.TextData(message));
+
+            string logo = "file:///" + Path.Combine(ProcessHelper.GetCurrentProcessLocation(), GVFSIconName);
+            binding.Items.Add(new BindingItem.ImageData()
+            {
+                Source = logo,
+                Placement = "appLogoOverride",
+                HintCrop = "circle"
+            });
+
+            if (!string.IsNullOrEmpty(actionButtonTitle))
+            {
+                ActionsData actionsData = new ActionsData();
+                actionsData.Actions = new XmlList<ActionItem>();
+                actionsData.Actions.Add(new ActionItem()
+                {
+                    Content = actionButtonTitle,
+                    Arguments = string.IsNullOrEmpty(callbackArgs) ? string.Empty : callbackArgs,
+                    ActivationType = "background"
+                });
+
+                toastData.Actions = actionsData;
+            }
+
+            XmlDocument toastXml = new XmlDocument();
+            using (StringWriter stringWriter = new StringWriter())
+            using (XmlWriter xmlWriter = XmlWriter.Create(stringWriter, new XmlWriterSettings { OmitXmlDeclaration = true }))
+            {
+                XmlSerializer serializer = new XmlSerializer(toastData.GetType());
+                XmlSerializerNamespaces namespaces = new XmlSerializerNamespaces();
+                namespaces.Add(string.Empty, string.Empty);
+
+                serializer.Serialize(xmlWriter, toastData, namespaces);
+
+                toastXml.LoadXml(stringWriter.ToString());
+            }
+
+            ToastNotification toastNotification = new ToastNotification(toastXml);
+            toastNotification.Activated += this.ToastActivated;
+            toastNotification.Dismissed += this.ToastDismissed;
+            toastNotification.Failed += this.ToastFailed;
+
+            ToastNotifier toastNotifier = ToastNotificationManager.CreateToastNotifier(this.appId);
+            toastNotifier.Show(toastNotification);
+        }
+
+        private void ToastActivated(ToastNotification sender, object e)
+        {
+            ToastActivatedEventArgs args = (ToastActivatedEventArgs)e;
+
+            this.UserResponseCallback?.Invoke(args.Arguments);

```

Comment:
Do we need a `try/catch` around this or in the callback method?

Reply:
Yes... I've added it around the `Process.Start` inside the callback. The other place that could throw exception is `TryValidatePath` which already has a `catch` block.